### PR TITLE
Update gogo/protobuf to v1.3.2 and k8s.io/* to v0.20.6 to fix CVE-2021-3121

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area security open-source
/kind enhancement
/priority 2

**What this PR does / why we need it**:
With this PR, [CVE-2021-3121](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2021-3121) is being fixed.
/cc @ThormaehlenFred 

**Special notes for your reviewer**:
There are four commits in this PR:
- (1) is similar to https://github.com/kubernetes/kubernetes/pull/98477, i.e., it vendors `gogo/protobuf@v1.3.2` and regenerates the protobuf code.
- (2) is a temporary workaround for https://github.com/kubernetes/kubernetes/pull/101306. Is was cherry-picked to the `release-v1.20` branch of k/k with https://github.com/kubernetes/kubernetes/pull/101326, however, it will only be released with `v1.20.7` which is planned for May, 12th (ref: https://groups.google.com/g/kubernetes-announce/c/Vixft2lPyaA/m/lGunIFjpAAAJ). Hence, for the time being, we need to manually apply the fix.
- (3) updates the `k8s.io/*` dependencies to `v0.20.6` to adapt the fix for the protobuf code of the Kubernetes API types.
- (4) drops the go.mod replacement directive for `sigs.k8s.io/structured-merge-diff/v4` as requested in https://github.com/gardener/gardener/pull/3945#pullrequestreview-644576714.

/squash

**Release note**:

```other operator
NONE
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -165,6 +165,8 @@ revendor:
 	@GO111MODULE=on go mod vendor
 	@GO111MODULE=on go mod tidy
 	@curl -sSLo hack/setup-envtest.sh https://raw.githubusercontent.com/kubernetes-sigs/controller-runtime/$(CR_VERSION)/hack/setup-envtest.sh
+# TODO: Remove this once k8s.io/apimachinery is upgraded to v0.20.7, see https://github.com/kubernetes/kubernetes/pull/101326
```
Comment (ialidzhikov) [inline Makefile:168]: It is strange that you apply such workaround for `resource.Quantity` type, but you didn't update the k/k dependencies to adopt the fix for all other types that we use. For example our types reference `corev1.ObjectReference`, `corev1.LocalObjectReference`, `corev1.Taint` and many others.
I suggest to update to k/k@v0.20.6 to adopt also the changes for the k/k types.

**Review by ialidzhikov — CHANGES_REQUESTED** [review]

**Review by ialidzhikov — COMMENTED** [review]
As part of the update to k/k@v0.20.6, the following replace can be now removed:

```diff
diff --git a/go.mod b/go.mod
index e8adece98..d6b484cfd 100644
--- a/go.mod
+++ b/go.mod
@@ -85,5 +85,4 @@ replace (
        k8s.io/helm => k8s.io/helm v2.13.1+incompatible
        k8s.io/kube-aggregator => k8s.io/kube-aggregator v0.20.6
        sigs.k8s.io/controller-runtime => github.com/gardener/controller-runtime v0.8.3-gardener.1
-       sigs.k8s.io/structured-merge-diff/v4 => sigs.k8s.io/structured-merge-diff/v4 v4.0.3 // remove after https://github.com/kubernetes/kubernetes/pull/99038 is merged and released
 )
```

**Comment by rfranzke** [comment]
@ialidzhikov Nice! Done

**Review by ialidzhikov — APPROVED** [review]
/lgtm

**Pull request merged**